core/src/main/java/io/questdb/cairo/CairoEngine.java

```diff
@@ -248,6 +282,97 @@ public IDGenerator getTableIdGenerator() {
         return tableIdGenerator;
     }
 
+    @Override
+    public TableWriterFrontend getTableWriterFrontEnd(
+            CairoSecurityContext securityContext,
+            CharSequence tableName,
+            @Nullable String lockReason
+    ) {
+        securityContext.checkWritePermission();
+        if (tableRegistry.hasSequencer(tableName)) {
+            return tableRegistry.getWalWriter(tableName);
+        }
+        return getWriter(securityContext, tableName, lockReason);
+    }
+
+    public void checkNotifyOutstandingTxnInWal(int tableId, CharSequence tableName, long txn) {
+        rootPath.trimTo(rootPathLen).concat(tableName).concat(TableUtils.TXN_FILE_NAME).$();
+        try (TxReader txReader = tempTxReader.ofRO(rootPath, PartitionBy.NONE)) {
+            if (txReader.unsafeLoad(true)) {
+                if (txReader.getTxn() < txn) {
+                    // table name should be immutable when in the notification
+                    String tableNameStr = Chars.toString(tableName);
+                    notifyWalTxnCommitted(tableId, tableNameStr, txn);
+                }
+            }
+        }
+    }
+
+    public void notifyWalTxnCommitted(int tableId, String tableName, long txn) {
+        Sequence pubSeq = messageBus.getWalTxnNotificationPubSequence();
+        int steelingAttempts = 100;
+        SqlToOperation sqlToOperation = null;
+        IntLongHashMap localCommittedTransactions = new IntLongHashMap();
+
+        try {
+            while (true) {
+                long cursor = pubSeq.next();
+                if (cursor > -1L) {
+                    WalTxnNotificationTask task = messageBus.getWalTxnNotificationQueue().get(cursor);
+                    task.of(tableName, tableId, txn);
+                    pubSeq.done(cursor);
+                    return;
+                } else if (cursor == -1L) {
+                    // Oh, no queue overflow!
+                    // Steel the work!
+                    if (steelingAttempts-- > 0) {
+                        Sequence subSeq = messageBus.getWalTxnNotificationSubSequence();
+                        if (sqlToOperation == null) {
+                            sqlToOperation = new SqlToOperation(this);
```

Comment:
This is very heavy handed, we need to solve creating one off `SqlCompiler` instances. `SqlCompiler` are too heavy objects from both GC and CPU point of view

Reply:
maybe we could have one per thread and reuse it.
another option is that caller has to pass it in so Sequencers and ApplyJobs would have their own instance.

Reply:
ApplyJob should have it, np, but ApplyJob does not use this method. Sequencer is thread safe, we cannot put stateful SqlCompiler there easily. 

Perhaps we can create thread safe pool of SqlCompilers on the engine level and use it everywhere (including current compilations).

Finally maybe we can wait here if we find a transaction which need SqlCompiler until notification slot appears instead of trying to process the queue.

Reply:
Added `ThreadSafeObjectPool` of SqlCompiler objects in `CairoEngine` to pool SqlCompiler instances.